tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuning.scala

```diff
@@ -0,0 +1,616 @@
+ *    Output:
+ *      Cannot recommend properties. See Comments.
+ *
+ *      Comments:
+ *      - java.io.FileNotFoundException: File worker_info.yaml does not exist
+ *      - 'spark.executor.memory' should be set to at least 2GB/core.
+ *      - 'spark.executor.instances' should be set to (gpuCount * numWorkers).
+ *      - 'spark.task.resource.gpu.amount' should be set to Max(1, (numCores / gpuCount)).
+ *      - 'spark.rapids.sql.concurrentGpuTasks' should be set between [1, 4]
+ *      - 'spark.rapids.memory.pinnedPool.size' should be set up to 4096m.
+ *      - 'spark.sql.adaptive.enabled' should be enabled for better performance.
+ */
+class AutoTuning(
+    val clusterProps: ClusterProperties,
+    val appInfo: Option[ApplicationSummaryInfo])  extends Logging {
+
+  import AutoTuning._
+
+  var comments: Seq[String] = Seq()
+  var recommendations: mutable.LinkedHashMap[String, RecommendationEntry] =
+    mutable.LinkedHashMap[String, RecommendationEntry]()
+
+  def getSparkPropertyFromProfile(propKey: String): Option[String] = {
+    if (appInfo.isDefined) {
+      appInfo.get.sparkProps.collectFirst {
+        case entry: RapidsPropertyProfileResult
+          if entry.key == propKey && entry.rows(1) != "null" => entry.rows(1)
+      }
+    } else {
+      None
+    }
+  }
+
+  def getPropertyValue(key: String): Option[String] = {
+    val fromProfile = getSparkPropertyFromProfile(key)
+    if (fromProfile.isDefined) {
+      fromProfile
+    } else {
+      Option(clusterProps.softwareProperties.get(key))
+    }
+  }
+
+  def initRecommendations(): Unit = {
+    recommendationsTarget.foreach { key =>
+      val propVal = getPropertyValue(key)
+      if (propVal.isDefined) {
+        val recommendationVal = new RecommendationEntry(key, propVal.get, null)
+        recommendations(key) = recommendationVal
+      }
+    }
+  }
+
+  def appendRecommendation(key: String, value: String): Unit = {
+    val recomRecord = recommendations.getOrElseUpdate(key,
+      new RecommendationEntry(key, null, null))
+    recomRecord.setRecommendedValue(value)
+    if (recomRecord.originalValue == null && value != null) {
+      // add a comment
+      appendComment(s"'$key' was not set.")
+    }
+  }
+
+  def appendRecommendationForMemoryMB(key: String, value: String): Unit = {
+    if (value != null) {
+      appendRecommendation(key, s"${value}m")
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.executor.instances' based on number of gpus and workers.
+   * Assumption - In case GPU properties are not available, it assumes 1 GPU per worker.
+   */
+  def calcExecInstances(): Int = {
+    // spark.executor.instances
+    clusterProps.gpu.getCount * clusterProps.system.numWorkers
+  }
+
+  /**
+   * Recommendation for 'spark.executor.cores' based on number of cpu cores and gpus.
+   */
+  def calcNumExecutorCores(): Int = {
+    // spark.executor.cores
+    val executorsPerNode = clusterProps.gpu.getCount
+    if (executorsPerNode == 0) {
```

Comment:
if this is zero either we didn't get the cluster properties or we really have zero, I guess we assume the former here? Do we want to append a comment?